Add TxInfo golden tests

teodanciu opened the pull request:
# Description
resolves https://github.com/input-output-hk/cardano-ledger/issues/3358

For Alonzo, Babbage and Conway:

* Adds an executable that generates a file containing cbor representation of arguments passed to `  ExtendedUTxO .txInfo` and the produced result. (For conway, running this executable is failing with `undefined`, due to the implementation of TxInfo in Conway)

For Alonzo and Babbage: 
* Adds a test that deserializes this file and for each entry, calls `ExtendedUTXO.txInfo` with the same arguments and expects the result to be the same.


# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [ ] The version bounds in `.cabal` files are updated
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Add an executable to generate a golden translations file

storing arbitrary arguments passed to `alonzoTxInfo` and the result.

teodanciu committed:
Add `GoldenTranslation`test checking that`alonzoTxInfo` returns expected

results

JaredCorduan commented on eras/alonzo/test-suite/test/Test/Cardano/Ledger/Alonzo/GoldenTranslation.hs:
```diff
@@ -0,0 +1,66 @@
+checkTranslationInstances = mapM_ checkTranslationInstance
+  where
+    checkTranslationInstance ti@(TranslationInstance pp utxo tx expected) =
+      let result = alonzoTxInfo pp PlutusV1 epochInfo systemStart utxo tx
```
maybe we should put the language type (eg `PlutusV1`) in `TranslationInstance` as well.

teodanciu replied:
Done

JaredCorduan commented:
@teodanciu this looks fantastic, exactly what we need!

teodanciu committed:
Parameterize `TxInfo` generation based on Plutus Language version

and store the language in the serialized TranslationInstance

teodanciu committed:
Generalize golden translation file generation and golden test

for any Alonzo+ era

- Introduce `TranslatableGen` class to parameterize how to generate
transactions and utxo that are sucessfully passed to `txInfo` per era

teodanciu committed:
Introduce golden translation file generation and test for Babbage

teodanciu committed:
Introduce golden translation file generation and test for Conway

JaredCorduan reviewed (approved):
@teodanciu this looks absolutely fantastic!

it looks like the CI failure is coming from:
```

<no location info>: error: [-Wunused-packages, -Werror=unused-packages]
    The following packages were specified via -package or -package-id flags,
    but were not needed for compilation:
      - cardano-ledger-alonzo-test-1.1.2.0
```

but I didn't see from where...

teodanciu commented:
> @teodanciu this looks absolutely fantastic!
> 
> it looks like the CI failure is coming from:
> 
> ```
> 
> <no location info>: error: [-Wunused-packages, -Werror=unused-packages]
>     The following packages were specified via -package or -package-id flags,
>     but were not needed for compilation:
>       - cardano-ledger-alonzo-test-1.1.2.0
> ```
> 
> but I didn't see from where...

@JaredCorduan This is so weird, only happens on ghc 8.10.7, which is why it didn't happen to me locally.
How...

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/Golden.hs:
```diff
@@ -0,0 +1,102 @@
+
+import Cardano.Ledger.Alonzo.TxInfo (ExtendedUTxO (..), VersionedTxInfo)
+import Cardano.Ledger.Binary (Annotator, DecCBOR (..))
+import Cardano.Ledger.Core -- (eraProtVerHigh, Era, PParams, PParamsHKD, Tx, TxOut)
```
```suggestion
import Cardano.Ledger.Core
```

lehins commented on eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Translation/GenerateGoldenFileMain.hs:
```diff
@@ -0,0 +1,77 @@
+  tgTx l = BabbageTranslatableGen.genTx @Conway l (genTxBody l)
+  tgUtxo = BabbageTranslatableGen.utxoWithTx @Conway
+
+genTxBody :: forall era. Crypto era => Language -> Gen (ConwayTxBody (ConwayEra era))
```
```suggestion
genTxBody :: forall c. Crypto c => Language -> Gen (ConwayTxBody (ConwayEra c))
```

teodanciu replied:
Changed also all the others `era`s that where defined like this to  `c`.

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/Golden.hs:
```diff
@@ -0,0 +1,102 @@
+  FilePath ->
+  IO ()
+generateGoldenFile ls file =
+  do
```
We should try to stick to a consistent style. Could you switch it in all other places as well?
```suggestion
generateGoldenFile ls file = do
```

teodanciu replied:
certainly

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/TranslationInstanceGen.hs:
```diff
@@ -0,0 +1,81 @@
+  [Language] ->
+  IO [TranslationInstance era]
+translationInstances size ls =
+  generate $ vectorOf size (genTranslationInstance ls)
```
It will be better if we have reproducable data generation, which `generate` can't provide. I don't think QuickCheck has a function that can do it for us, but we can write one:
```suggestion
  Int ->
  [TranslationInstance era]
translationInstances size ls seed =
  generateWithSeed seed $ vectorOf size (genTranslationInstance ls)
  
generateWithSeed :: Int -> Gen a -> a
generateWithSeed seed (MkGen g) = g (mkQCGen seed) 30
```

Requires import of `Test.QuickCheck.Random` and `Test.QuickCheck.Gen`

A random seed can just be hard coded.

teodanciu replied:
That cool! Thanks 
I verified that re-running in this version doesn't change the file .

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/TranslationInstanceGen.hs:
```diff
@@ -0,0 +1,81 @@
+epochInfo = fixedEpochInfo (EpochSize 100) (mkSlotLength 1)
+
+systemStart :: SystemStart
+systemStart = SystemStart $ posixSecondsToUTCTime 0
```
Something more realistic
```suggestion
systemStart = SystemStart $ posixSecondsToUTCTime 1684365002711
```

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/TranslationInstanceGen.hs:
```diff
@@ -0,0 +1,81 @@
+  Gen (TranslationInstance era)
+genTranslationInstance ls = do
+  pp <- arbitrary :: Gen (PParams era)
+  language <- oneof (pure <$> ls)
```
```suggestion
  language <- elements ls
```

lehins commented on eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal:
```diff
@@ -65,17 +69,65 @@ library
 
+executable gen-golden
+    main-is:
+        Test/Cardano/Ledger/Alonzo/Translation/GenerateGoldenFileMain.hs
```
Main file is usually never buried that deep:
```suggestion
        GenerateGoldenFileMain.hs
```

lehins commented on eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal:
```diff
@@ -65,17 +69,65 @@ library
+    main-is:
+        Test/Cardano/Ledger/Alonzo/Translation/GenerateGoldenFileMain.hs
+
+    hs-source-dirs:   src
```
And `main` should never be exported from a library
```suggestion
    hs-source-dirs:   test
```

lehins commented on eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal:
```diff
@@ -65,17 +69,65 @@ library
+        Test.Cardano.Ledger.Alonzo.Serialisation.Generators
+        Test.Cardano.Ledger.Alonzo.Translation.TranslationInstance
+        Test.Cardano.Ledger.Alonzo.Translation.TranslationInstanceGen
+        Test.Cardano.Ledger.Alonzo.Translation.Golden
```
It is not a good approach to reuse modules liked that. That's what the library is for. Listing same modules in different section will lead to them being recompiled twice.

So, just depend on `cardano-ledger-alonzo-test` instead
```suggestion
```

lehins commented on eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal:
```diff
@@ -99,7 +151,7 @@ test-suite cardano-ledger-alonzo-test
         cardano-ledger-alonzo-test,
         cardano-ledger-binary:{cardano-ledger-binary, testlib},
-        cardano-ledger-core,
+        cardano-ledger-core >=1.0,
```
No need to specify bounds, the library section already sets the bounds.
```suggestion
        cardano-ledger-core,
```

lehins commented on eras/alonzo/test-suite/golden/translations.cbor:
Make sure to list data files in cabal file in `data-files:` section

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/Golden.hs:
```diff
@@ -0,0 +1,102 @@
+        <> "\n utxo: "
+        <> show utxo
+        <> "\n tx: "
+        <> show tx
```
A bit easier to read and more compact:
```suggestion
         errorMessage = unlines
           [ "Unexpected txinfo with arguments: "
           , " pp: " <> show pp
           , " language: " <> show l
           , " utxo: " <> show utxo
           , " tx: " <> show tx
           ]
       in case result of
            Left e -> error $ show e
            Right info -> TxInfoResultComparison expected info errorMessage
```

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/GenerateGoldenFileMain.hs:
```diff
@@ -0,0 +1,34 @@
+  tgUtxo _ tx = do
+    let ins = tx ^. bodyTxL ^. inputsTxBodyL
+    outs <- vectorOf (length ins) (arbitrary :: Gen (TxOut Alonzo))
+    pure $ UTxO (Map.fromList $ Set.toList ins `zip` outs)
```
Why not put this instance into `TranslateableGen` module, like in later eras?

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/TranslationInstanceGen.hs:
```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Test.Cardano.Ledger.Alonzo.Translation.TranslationInstanceGen (
```
For consistency sake, let's call it:
```suggestion
module Test.Cardano.Ledger.Alonzo.Translation.TranslatableGen (
```

teodanciu replied:
Yes, perfect!

lehins commented on eras/conway/test-suite/src/Test/Cardano/Ledger/Conway/Translation/GenerateGoldenFileMain.hs:
```diff
@@ -0,0 +1,77 @@
+main :: IO ()
+main = generateGoldenFile @Conway [PlutusV1, PlutusV2, PlutusV3] "eras/conway/test-suite/golden/translations.cbor"
+
+instance TranslatableGen Conway where
```
Let's keep the `main` file minimal and move the instance into `TranslatableGen` module

It will be easier to find them in the future.

lehins commented on eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/Translation/TranslatableGen.hs:
```diff
@@ -0,0 +1,163 @@
+  let ins = tx ^. bodyTxL ^. inputsTxBodyL
+  let refIns = tx ^. bodyTxL ^. referenceInputsTxBodyL
+  let allIns = ins `Set.union` refIns
+  outs <- vectorOf (length allIns) (genTxOut @era l)
```
```suggestion
  let allIns = tx ^. bodyTxL ^. allInputsTxBodyL
  outs <- vectorOf (length allIns) (genTxOut @era l)
```

teodanciu replied:
Nice, this allowed me to get rid of this constraint : `BabbageEraTxBody era`

lehins commented on eras/alonzo/test-suite/test/Test/Cardano/Ledger/Alonzo/GoldenTranslation.hs:
```diff
@@ -0,0 +1,46 @@
+    [testCase "golden/translations.cbor" $ check "golden/translations.cbor"]
+
+check :: String -> Assertion
+check file = do
```
```suggestion
    [testCase "golden/translations.cbor" $ goldenAssertion "golden/translations.cbor"]

goldenAssertion :: FilePath -> Assertion
goldenAssertion file = do
```

lehins commented on eras/babbage/test-suite/cardano-ledger-babbage-test.cabal:
```diff
@@ -56,13 +58,45 @@ library
+    other-modules:
+        Paths_cardano_ledger_babbage_test
+        Test.Cardano.Ledger.Babbage.Serialisation.Generators
+        Test.Cardano.Ledger.Babbage.Translation.TranslatableGen
```
ditto
```suggestion
```

lehins commented on eras/babbage/test-suite/cardano-ledger-babbage-test.cabal:
```diff
@@ -56,13 +58,45 @@ library
+    main-is:
+        Test/Cardano/Ledger/Babbage/Translation/GenerateGoldenFileMain.hs
+
+    hs-source-dirs:   src
```
Ditto
```suggestion
    hs-source-dirs:   test
```

lehins commented on eras/babbage/test-suite/cardano-ledger-babbage-test.cabal:
```diff
@@ -56,13 +58,45 @@ library
 
+executable gen-golden
+    main-is:
+        Test/Cardano/Ledger/Babbage/Translation/GenerateGoldenFileMain.hs
```
```suggestion
        GenerateGoldenFileMain.hs
```

lehins commented on eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/Translation/TranslatableGen.hs:
```diff
@@ -0,0 +1,163 @@
+    pure $
+      arbWits
+        & rdmrsTxWitsL @era
+          .~ Redeemers (Map.singleton (RdmrPtr Spend 0) (Data (PV1.I 42), ExUnits 5000 5000))
```
This looks odd. How come a single `Redeemer` is hard coded for PlutusV2 and PlutusV3?

teodanciu replied:
Mhm. It's because in v1 this is not checked, so I chose the easiest way to make the transaction validate. 
But it's perhaps too lazy of me, let me try to generate data instead of this

teodanciu replied:
I improved this slightly, with the least amount of hardcoding that allows us to not add more complexity to the generation of the txbody. This is because the redeemers have a relationship with the contents of the body. 
Since this is validated in the rule anyway, I thought it's not important to be so accurate about generating data for these tests.

lehins reviewed (changes requested):
This looks really great. Some things need to be moved around, but otherwise this is exactly what we need.

teodanciu committed:
Improve based on feedback

- move `TranslatableGen` class to it's own module and move logic for
generating `TranslationInstance` there (previously
TranslationInstanceGen)
- move main classes to `test` folder, in order to not expose them in the
library
- fix cabal of executables: depend on the test library instead of
declaring dependencies in `other-modules`

teodanciu committed:
Generate random `TranslationInstance` data using a seed

so that the generation is reproducible

teodanciu committed:
Improve generation of Redeemers in `TranslatableGen`

lehins commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/Golden.hs:
```diff
@@ -0,0 +1,97 @@
+compareTxInfoResults ::
+  forall era.
+  ( EraTx era
+  , ExtendedUTxO era
```
It would be useful to add stack trace to such helper function, otherwise it won't be clear where the error actually originated
```suggestion
  , ExtendedUTxO era
  , HasCallStack
```

lehins commented on eras/babbage/test-suite/test/Test/Cardano/Ledger/Babbage/GoldenTranslation.hs:
```diff
@@ -0,0 +1,46 @@
+    ( \(TxInfoResultComparison expected actual err) ->
+        assertEqual err expected actual
+    )
+    comps
```
That's the same function as in `Test.Cardano.Ledger.Alonzo.GoldenTranslation.goldenAssertion`

We should be able to parametrize it on `era`, put into a `Test.Cardano.Ledger.Alonzo.Translation.Golden` and reuse it here, right?

teodanciu replied:
Yes, I was under the false impression that we wanted to avoid using tasty (and friends: hunit etc) in the exposed library. I don't know why I thought this, but since it's not true, I will remove the whole `TxInfoResultComparison`,  and just create a function as you suggested parameterized per era, that asserts the equality.

lehins reviewed (approved):
Looks great. Just a couple of minor suggestions that cought my eye on a follow up review.

teodanciu committed:
Simplify assertion of translation equality with golden file records

Also add `HasCallStack` to make the origin of a failure easier to
understand.
